qcqplib/Solver.cpp

```diff
@@ -0,0 +1,414 @@
+#include <iostream>
+#include <Eigen/Dense>
+#include <chrono>
+#include <vector>
+#include "Solver.hpp"
+
+using namespace std;
+using namespace Eigen;
+
+Solver::Solver(){
+    prob_id = 1;
+}
+
+VectorXd Solver::iterative_refinement(MatrixXd A, VectorXd b, double mu_ir = 1e-7, double epsilon = 1e-10, int max_iter = 10){
+    VectorXd Ab(A.cols()), delta(A.cols());
+    MatrixXd AA_tild(A.cols(), A.cols()),AA_tild_inv(A.cols(), A.cols());
+    VectorXd x = VectorXd::Zero(A.cols());
+    Ab = A.transpose()*b;
+    //MatrixXd A_t = A.transpose();
+    AA_tild = A.transpose()*A;
+    MatrixXd Id = MatrixXd::Identity(AA_tild.rows(),AA_tild.cols());
+    AA_tild_inv = AA_tild+mu_ir*Id;
+    //SelfAdjointEigenSolver<MatrixXd> eigensolver(AA_tild);
+    //if (eigensolver.info() != Success) abort();
+    //std::cout << " A_t  : " << A_t << std::endl;
+    //std::cout << " A  : " << A << std::endl;
+    //std::cout << " dd  : " << b << std::endl;
+    //cout << "The eigenvalues of AA are:\n" << eigensolver.eigenvalues() << endl;
+    //AA_tild_inv = AA_tild_inv.inverse();
+    AA_tild_inv = AA_tild_inv.llt().solve(Id);
```

Comment:
There is memory allocation here.
```suggestion
    AA_tild_inv.setIdentity(); AA_tild.llt().solveInPlace(AA_tild_inv);
```